micro/example_test.go

```diff
@@ -43,56 +47,81 @@ func ExampleAddService() {
 		},
 
 		// DoneHandler can be set to customize behavior on stopping a service.
-		DoneHandler: func(srv micro.Service) {
-			info := srv.Info()
-			fmt.Printf("stopped service %q with ID %q\n", info.Name, info.ID)
+		DoneHandler: func(ctx context.Context, srv micro.Service) {
+			info := srv.Info(ctx)
+			logger := slog.FromContext(ctx)
+			logger.Info(
+				"stopped service",
+				slog.String("name", info.Name),
+				slog.String("id", info.ID),
+			)
 		},
 
 		// ErrorHandler can be used to customize behavior on service execution error.
-		ErrorHandler: func(srv micro.Service, err *micro.NATSError) {
-			info := srv.Info()
-			fmt.Printf("Service %q returned an error on subject %q: %s", info.Name, err.Subject, err.Description)
+		ErrorHandler: func(ctx context.Context, srv micro.Service, err *micro.NATSError) {
+			info := srv.Info(ctx)
+			logger := slog.FromContext(ctx)
+			logger.Error(
+				"service error",
+				err,
+				slog.String("name", info.Name),
+				slog.String("subject", err.Subject),
+				slog.String("description", err.Description),
+			)
 		},
 	}
 
-	srv, err := micro.AddService(nc, config)
+	logger := slog.FromContext(ctx)
+	srv, err := micro.AddService(ctx, nc, config)
 	if err != nil {
-		log.Fatal(err)
+		logger.Error("failed to add service", err)
+		os.Exit(1)
 	}
-	defer srv.Stop()
+	defer srv.Stop(ctx)
 }
 
 func ExampleService_Info() {
+	ctx := exampleCtx()
+	logger := slog.FromContext(ctx)
+
 	nc, err := nats.Connect("127.0.0.1:4222")
 	if err != nil {
-		log.Fatal(err)
+		logger.Error("failed to connect to NATS", err)
+		os.Exit(1)
 	}
 	defer nc.Close()
 
 	config := micro.Config{
 		Name: "EchoService",
 		Endpoint: micro.Endpoint{
 			Subject: "echo",
-			Handler: micro.HandlerFunc(func(micro.Request) {}),
+			Handler: micro.HandlerFunc(func(context.Context, micro.Request) {}),
```

Comment:
I don't think it might be necessary to be changing these, what could be done is take advantage of the interface func here and have a wrapper for the handler, maybe something like:

```go 
func WithLoggingContext(ctx context.Context, h micro.Handler) micro.Handler {
	return micro.HandlerFunc(func(req micro.Request) {
             logger := slog.FromContext(ctx) // do context things
             logger.Info(...)
             h(req)
	})
}
```

What I do see is that you want to be able to inspect whether the response was successful or not after responding, maybe there could be a way to inspect via the interface whether the internal `respondError` for the request.

Reply:
Good point, happy to change. I wasn't sure about the details of how it should be handled, just didn't want to make an issue with at least putting some effort into a version that passes the current tests.

Reply:
How does this help if you need to use the context from within the handler?

Reply:
Something like this?
```go
func WithContext(h func(context.Context, micro.Request)) micro.Handler {
	return micro.HandlerFunc(func(req micro.Request) {
		h(ctx, req)
	})
}
```

Reply:
You could do something like this instead:
```go
func WithContext(ctx context.Context, h func(context.Context, micro.Request)) micro.Handler {
	return micro.HandlerFunc(func(req micro.Request) {
		h(ctx, req)
	})
}

func ExampleAddService_Custom() {
	nc, err := nats.Connect("127.0.0.1:4222")
	if err != nil {
		log.Fatal(err)
	}
	defer nc.Close()

	echoHandler := func(ctx context.Context, req micro.Request) {
		log.Println("With Context:", ctx.Value("logging"))
		req.Respond(req.Data())
	}

	ctx := context.Background()
	ctx = context.WithValue(ctx, "logging", "example")

	config := micro.Config{
		Name:        "EchoService",
		Version:     "1.0.0",
		Description: "Send back what you receive",
		Endpoint: micro.Endpoint{
			Subject: "echo",
			Handler: WithContext(ctx, echoHandler),
		},
}
```

Reply:
Agreed, I think leveraging `Handler` interface to pass basically anything you want to a handler (context, just logger etc.) would be the way to go here. 

I was wondering whether we could mimic what HTTP does with context embedded in `http.Request` and then retrieving it using `req.Context()`, but I am not convinced (as @eventgraph mentioned, that does not sit well with how you would use context in go).

Reply:
@piotrpio we don't really care where it lives, just that I have a context to work with.  Context as the first argument is the norm, the addition on `http.Request` I think was the only way to keep the interface the same without breakage as Context didn't come about for a while.  Since this is a new alpha contract my hope is you can fix it to be "right".

Reply:
Would you like me to close this PR and make a new Issue / Feature Request?

Reply:
Hey, sorry for the delayed response. I created an issue (https://github.com/nats-io/nats.go/issues/1196) and will implement soon. I'm closing this PR for now.